rust/worker/src/execution/operators/prefetch_keys.rs

```diff
@@ -0,0 +1,107 @@
+use async_trait::async_trait;
+use chroma_blockstore::{
+    arrow::types::{ArrowReadableKey, ArrowReadableValue},
+    key::{InvalidKeyConversion, KeyWrapper},
+    memory::Readable,
+    provider::{BlockfileProvider, OpenError},
+    BlockfileReader, Key, Value,
+};
+use chroma_error::{ChromaError, ErrorCodes};
+use chroma_system::{Operator, OperatorType};
+use futures::FutureExt;
+use parking_lot::Mutex;
+use std::{future::Future, pin::Pin};
+use thiserror::Error;
+use uuid::Uuid;
+
+#[derive(Error, Debug)]
+pub enum PrefetchKeysError {
+    #[error("Error constructing reader: {0}")]
+    CouldNotConstructReader(#[from] Box<OpenError>),
+}
+
+impl ChromaError for PrefetchKeysError {
+    fn code(&self) -> ErrorCodes {
+        match self {
+            PrefetchKeysError::CouldNotConstructReader(e) => e.code(),
+        }
+    }
+}
+
+pub struct PrefetchKeysInput {
+    #[allow(clippy::type_complexity)]
+    prefetch_fut:
+        Mutex<Option<Pin<Box<dyn Future<Output = Result<(), PrefetchKeysError>> + Send>>>>,
+}
+
+impl std::fmt::Debug for PrefetchKeysInput {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("PrefetchKeysInput").finish()
+    }
+}
+
+impl PrefetchKeysInput {
+    pub fn new<
+        'referred_data,
+        K: Key
+            + Into<KeyWrapper>
+            + TryFrom<&'referred_data KeyWrapper, Error = InvalidKeyConversion>
+            + ArrowReadableKey<'referred_data>
+            + Sync
+            + Send
+            + 'referred_data + 'static,
+        V: Value
+            + Readable<'referred_data>
+            + ArrowReadableValue<'referred_data>
+            + Sync
+            + Send
+            + 'referred_data,
```

Comment:
I hate these generics. It's annoying because `load_blocks_for_keys()` actually only needs `K: Key`, but we need a full reader to get at it.